Update computing pillar

jemrobinson opened the pull request:
## :white_check_mark: Checklist

<!--
Replace the empty checkboxes [ ] below with checked ones [x] accordingly.
-->

- [x] This pull request has a meaningful title.
- [x] If your changes are not yet ready to merge, you have marked this pull request as a **draft** pull request.

## :ballot_box_with_check: Maintainers' checklist

<!--
This checklist is for project maintainers to use after the pull request is submitted.
Feel free to leave these empty.
-->

- [x] This pull request has had the appropriate labels assigned
- [x] This pull request has been added to the SATRE backlog project board
- [x] This pull request has been assigned to one or more maintainers

### :arrow_heading_up: Summary

<!--
Please explain what your pull request does here.
-->

- Updates to the computing pillar following discussions with @machintim 
- Initial changes were a reordering of existing statements into updated capabilities (d8eec05)
- Some statements were added or combined (f5215f9)
- Other commits are (possibly) controversial removals which we can consider individuallly
  - [x] Your TRE should take accessibility for users with disabilities into account b02291c
  - [x] Your TRE could integrate with cloud-native managed services. 1988a36
  - [ ] ~Your TRE must provide software applications that are relevant a08505c~ reverted by be1f53a
  - [ ] ~Your TRE should be able to provide access to accelerators 38d7785~ reverted by 72140c2
  - [ ] ~Your TRE must maintain segregation of users and data 826ff45~ reverted by 8908397
  - [ ] ~You must have a documented procedure for making changes to deployed infrastructure 546e5d0~ reverted by a942cc6
  - [ ] ~You must monitor the network configuration of your TRE 9e8702b~ reverted by 39cbbe0
  - [ ] ~Infrastructure Analytics d3cb03f~ reverted by 75baaf5
 
### :closed_umbrella: Related issues

<!--
If your pull request will close any open issues add `Closes #<issue number>` here.
Also not any issues your pull request relates to, for example `Contributes to #<issue number>`.
-->

### :raising_hand: Acknowledging contributors

<!-- Please tick one of these boxes and list any contributors who should be recognised.-->

- [ ] All contributors to this pull request are already named in the [table of contributors](https://github.com/sa-tre/satre-specification#contributors) in the README file.
- [ ] The following people should be added to the [table of contributors](https://github.com/sa-tre/satre-specification#contributors) in the README file: <!-- replace this text with the GitHub IDs of any new contributors -->

JimMadge commented on docs/source/pillars/computing_technology.md:
```diff
@@ -104,59 +132,90 @@ _The ability to run analyses requiring more compute resources, or more specialis
+  - Importance
+* - You should understand the availability and uptime guarantees of any providers that you rely on.
+  - For remote TREs this might include your cloud provider(s) and/or data centre operators.
+    For on-premises TREs, it might be worth considering your ISP and electricity provider.
```
Is the electricity provider suggestion based on experience?

Definitely not my area but I'm not aware of uptime guarantees from utilities. I would think your primary concern here is that you have enough UPS capacity to safely shut everything down if there is a power cut.

jemrobinson replied:
Yeah, this seems weird to me. Feels like the kind of thing you should mitigate (eg. with UPS as you suggest) rather than obtain a third-party guarantee. Can you reword to make this point?

jemrobinson replied:
```suggestion
    For on-premises TREs, it might be worth using an uninterruptable power supply (UPS) and planning how you would deal with internet outages.
```

JimMadge commented on docs/source/pillars/computing_technology.md:
```diff
@@ -104,59 +132,90 @@ _The ability to run analyses requiring more compute resources, or more specialis
+
+### Network management
+
+An application used to manage and control network devices, ensuring proper functioning, security, and performance of the network.
```
```suggestion
Managing network infrastructure, ensuring proper functioning, security, and performance.
```

jemrobinson replied:
I think @machintim wants to make it clear which things are processes/applications/data assets. Can you reword in a way that isn't ambiguous?

jemrobinson replied:
```suggestion
An application used to manage network infrastructure, ensuring proper functioning, security, and performance.
```

JimMadge commented on docs/source/pillars/computing_technology.md:
```diff
@@ -165,9 +224,6 @@ _The ability of the TRE operator to administer and secure network infrastructure
   - Network infrastructure must prevent unauthorised access to resources on the network.
     This may include firewalls, network segmentation, and restricting connections to the network.
   - Mandatory
-* - You must monitor the network configuration of your TRE to check for misconfigurations and vulnerabilities.
-  - This may include regular vulnerability scanning, and penetration testing.
-  - Mandatory
```
This feels like an odd one to remove. How can we be confident a TRE is correctly configured, and secure, if we don't test that?

I think this covers more than just networking though.

jemrobinson replied:
I agree.

jemrobinson replied:
Reverted in 39cbbe0

JimMadge commented on docs/source/pillars/computing_technology.md:
````diff
@@ -176,79 +232,57 @@ _The ability of the TRE operator to administer and secure network infrastructure
-  - Mandatory
 ```
 
 ### Configuration management
````
This section feels in the wrong place. Shouldn't it be under (or alongside) infrastructure management?

jemrobinson replied:
The idea here is that billing is part of managing capacity. The only statement we had about billing was to do with infrastructure, but we could/should add something about personpower here too. Would that make it more coherent for you @JimMadge ?

jemrobinson replied:
One to raise with @machintim when he gets back.

JimMadge reviewed (changes requested):
The structure feels clearer.

I think some of the removals are questionable, or I have missed where they are covered elsewhere.

I don't like some of the sentences. I think it is often trying to be more specific than it needs to be which ends up with it being incorrect. Like saying network management or VCS is "an application" or lists which seem like they are mutual exclusive and collectively exhaustive but are not.

edwardchalstrey1 commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,55 @@
 
-_The interfaces used for interacting with the TRE management system and the TRE workspace._
+Software or system that allows people to interact with computers, applications, and services.
+This application group includes desktop, command line and code submission interface.
```
*interfaces* plural?

craddm commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,55 @@
 
+This pillar concerns actions taken by the TRE operator to manage systems for storing, retrieving, analysing and sending information.
+
+Each TRE operator will have its own computing technology requirements.
+Requirements are drawn from information governance requirements which primarily specify the controls needed within the technical and computing infrastructure.
+Researcher personas influence the according to the technical knowledge and experience of researchers along with the work they need to perform within the system, a data scientist will have very different requirements to a clinician for example.
```
I think this paragraph might be aiming for something like this?

```suggestion
Each TRE operator will have its own computing technology requirements.
Information governance requirements primarily specify security controls needed within the technical and computing infrastructure.
Other computing technology requirements will be influenced by the technical knowledge and experience of those using the TRE, along with the work they need to perform within the system. 
For example, a data scientist will have very different requirements to a clinician.
```

jemrobinson replied:
Are you happy with this @JimMadge ?

JimMadge replied:
Sounds good to me :+1:.

Might just be tempted to remove some of the adjectives for brevity.

edwardchalstrey1 commented on docs/source/pillars/computing_technology.md:
```diff
@@ -59,39 +59,67 @@ _The tools used by researchers inside a TRE, such as programming languages, IDEs
 * - Statement
   - Guidance
   - Importance
-* - Your TRE must provide software applications that are relevant to working with the data in the TRE.
-  - The tools provided will depend on the types of data in the TRE, and the expectations of users of the TRE.
-    For users working in a TRE via a virtual desktop, this may include programming languages such as Python and R, integrated development environments, Jupyter notebooks, office type applications such as word processors and spreadsheets, command line tools, etc.
-    TREs with non-desktop interfaces should similarly consider carefully which applications are best suited for the researchers needs when interacting with the data, for example "point and click" GUI tools for querying a database and generating plots of data.
-    The set of tools should be reviewed regularly to ensure they are up to date.
-  - Mandatory
```
Does the specification need an appendix where we could put examples of TRE components? This could read something like:

- Your TRE must provide software applications that are relevant to working with the data in the TRE.
- The set of tools provided will depend on the types of data in the TRE, and the expectations of the TRE users, and should be reviewed regularly to ensure they are up to date. For examples of the kinds of software applications used by existing TREs, refer to the [appendix] <- *link to relevant part*
- Mandatory

jemrobinson replied:
Where things have been removed, I'm not completely sure why. I think it's fine to say that we choose to keep them (and amend the WP3 document accordingly). @JimMadge: do you want to reword this to be a bit less verbose?

jemrobinson replied:
Reverted in be1f53a

edwardchalstrey1 commented on docs/source/pillars/computing_technology.md:
They aren't the same thing, but I do think here the sentiment is that version control helps with reproducible data analyses.

I think this section is good but I agree with Jim that the header should probably be renamed

jemrobinson replied:
If you take a look at the diagram in WP3 the point here is that a VCS system is an Application Component of the End-User-Computing capability. We need to say something sensible about **why** a VCS system is a useful component, but it doesn't have to be this. Thoughts @JimMadge/@edwardchalstrey1?

jemrobinson replied:
I'm going to ignore this to stay in-sync with WP3, but please comment on the doc in teams or discuss with @machintim when he gets back.

edwardchalstrey1 commented on docs/source/pillars/computing_technology.md:
```suggestion
### Enabling Reproducible Data Analysis and Version Control of Research Code
```

jemrobinson replied:
I don't love the current name but I think the point is that it's an "application" in architecture-language so the name needs to be a description of an object/system.

edwardchalstrey1 replied:
How about that? Feel free to reject entirely

jemrobinson replied:
It's not wrong, but it's a bit verbose! :joy:

edwardchalstrey1 replied:
how about now?

jemrobinson replied:
I'm going to ignore this to stay in-sync with WP3, but please comment on the doc in teams or discuss with @machintim when he gets back.

edwardchalstrey1 commented on docs/source/pillars/computing_technology.md:
```diff
@@ -176,79 +232,57 @@ _The ability of the TRE operator to administer and secure network infrastructure
-  - Optional
-* - You must have a documented procedure for removing infrastructure when it is no longer needed
-  - Removing unused infrastructure not only reduces costs and management burden but also reduces the attack surface of a TRE and reduces the risk of unaddressed vulnerabilities.
+* - You must have a procedure for increasing/decreasing available resources.
```
```suggestion
* - TRES with the ability to increase/decrease available resources must have a documented procedure in place for how to do this.
```

jemrobinson replied:
A bit of a different tack, but after re-reading the guidance, I think the following might be what we're trying to say here:

```suggestion
* - You must have a procedure for allocating available resources among projects.
```

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,55 @@
+
+Each TRE operator will have its own computing technology requirements.
+Requirements are drawn from information governance requirements which primarily specify the controls needed within the technical and computing infrastructure.
+Researcher personas influence the according to the technical knowledge and experience of researchers along with the work they need to perform within the system, a data scientist will have very different requirements to a clinician for example.
```
I think the essential point is that requirements for this pillar come from: IG, what the TRE organisation can offer and the needs of researchers. Do you want to suggest a better phrasing?

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,55 @@
-  - TREs which allow users to connect from their own devices should not require the installation of any bespoke TRE application on the user's device.
-    In practice a web browser is the most common way to achieve this.
+
+* - You should disable the ability to copy data out of your TRE via the system clipboard.
+  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
+    A TRE may allow user to paste text into a workspace.
   - Recommended
```
Feels like a "Mandatory" to me. What do you think?

JimMadge replied:
I would say so. It feels like it is a big risk for not much benefit. Maybe TREs who are much more confident about safe people are less concerned here. It might be that we came to recommended here to not be too prescriptive.

In any case, I don't think we can say "should not" in the statement and then "must not" in the guidance.

@manics do you remember any extra background context here?

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,55 @@
   - This may take the form of a virtual Windows or Linux desktops, non-desktop interfaces such as JupyterLab and other web applications, or a terminal.
     Bespoke TRE-specific software should be avoided when widely used alternatives already exist.
   - Recommended
-* - Your TRE should take accessibility for users with disabilities into account.
-  - The restricted nature of TREs means many assistive tools such as screenreaders in a virtual desktop may not be allowed, but other options such as colour schemes, font sizes, and resizing user interface elements, should be supported.
-  - Recommended
```
I think @machintim's suggestion is that this should be a SATRE-level principle that applies across everything, not a sentence under a single capability.

jemrobinson replied:
Moved into principles. @harisood to rework these based on recent Collaboration Cafe.

jemrobinson commented on docs/source/pillars/computing_technology.md:
Again, a non-ideal name for the Application Component that provides access to external software. What would you call it @JimMadge ?

jemrobinson commented on docs/source/pillars/computing_technology.md:
Agree, these could be merged. Can you make a suggestion to do so?

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -104,59 +132,90 @@ _The ability to run analyses requiring more compute resources, or more specialis
     For example this may be in the form of a batch scheduler on a HPC cluster, or a dynamically created compute nodes on a cloud platform.
   - Recommended
-* - Your TRE should be able to provide access to accelerators such as GPUs if required by users.
-  - GPUs and other accelerators are commonly used in machine learning and other computationally intensive research.
-    TREs should make it clear to users whether GPUs and other resources are available whilst projects are being assessed.
-  - Recommended
```
This is one of the things to discuss. I'm also in favour of keeping it.

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -104,59 +132,90 @@ _The ability to run analyses requiring more compute resources, or more specialis
+The ability of the TRE operator to instantiate, deploy, change or remove physical or virtual infrastructure.
 
-_The ability of the TRE operator to record and analyse data about the usage of the TRE._
+### Infrastructure deployment
+
+The process of setting up and configuring infrastructure components and resources to support applications or services.
+This involves development, installation, configuration, and validation.
```
Are those different Processes? Current processes here are:

- Infrastructure Deployment
- Change process
- Infrastructure Decommissing
- Availability Management

How would you prefer to split this?

jemrobinson commented on docs/source/pillars/computing_technology.md:
```diff
@@ -2,55 +2,56 @@
-  - TREs which allow users to connect from their own devices should not require the installation of any bespoke TRE application on the user's device.
-    In practice a web browser is the most common way to achieve this.
+
+* - You should disable the ability to copy data out of your TRE via the system clipboard.
+  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
+    A TRE may allow user to paste text into a workspace.
   - Recommended
```
```suggestion
* - You must not allow users to copy data out of your TRE via the system clipboard.
  - A TRE user must not be able to copy sensitive data out of a workspace using the system clipboard.
    A TRE may allow user to paste text into a workspace.
    This might not be relevant to your TRE, for example if your user interface does not have a clipboard.
  - Required
```

jemrobinson commented:
@JimMadge @craddm @edwardchalstrey1 . I've reverted the more controversial changes (and added notes to @machintim's WP3 document). Are there further structural changes you'd like to see?

JimMadge commented on docs/source/specification.md:
```diff
@@ -66,6 +66,7 @@ In addition, we also describe some {ref}`roles <satre_roles>` that are necessary
 The SATRE specification has been developed based on the following principles:
 
 - TREs should be as easy as possible for end-users to use (_e.g._ researchers) whilst still remaining secure.
+- TREs should take accessibility for all users with disabilities into account
```
```suggestion
- TREs should take accessibility for all users, including those with disabilities, into account.
```

Accessibility matters for all users. I suppose there is an argument that, for example, only providing a command-line interface for medical professionals is poor accessibility.

Do we want to say it is _particularly_ important to consider users with disabilities?

There might be TPS people at the Turing to would know how to better phrase this.

JimMadge reviewed (changes requested):
I think this is looking better.

A couple of suggestions, and I would move configuration management somewhere else.

One thing I still don't like (but it feels like we can't quite solve here) is using terms like "Application" and "Service" as elements of the specification. We are quite likely to use those words to mean very different things in the guidance. _E.g._ desktop application or HPC service. It could be quite confusing to readers (it was to me). Do we have a plan to make this clear, to define those terms somewhere?

edwardchalstrey1 commented:
> @JimMadge @craddm @edwardchalstrey1 @manics . I've reverted the more controversial changes (and added notes to @machintim's WP3 document). Are there further structural changes you'd like to see?

No further comments from me

jemrobinson commented:
> One thing I still don't like (but it feels like we can't quite solve here) is using terms like "Application" and "Service" as elements of the specification. We are quite likely to use those words to mean very different things in the guidance. E.g. desktop application or HPC service. It could be quite confusing to readers (it was to me). Do we have a plan to make this clear, to define those terms somewhere?

There are definitions in WP3. I'll reproduce them here:

- Business Process
  - A set of actions which produce a specific desired outcome. Example: To access the TRE a researcher needs to complete an onboarding business process.
- Application Component
  - An encapsulation of application functionality which is modular and replaceable.  Example: To perform work within a TRE a researcher might need access to a Desktop or command line interface application component.
- Data Object
  - A store of data or information. Example: To know what data is stored within the TRE a study database data object is needed. This contains information on the data assets within the TRE, who owns them and other compliance information.

harisood commented on docs/source/pillars/computing_technology.md:
```diff
@@ -100,6 +135,12 @@ _The ability to run analyses requiring more compute resources, or more specialis
+* - Your TRE must maintain segregation of users and data from different projects when using non-standard compute.
+  - High performance or specialist compute is often shared amongst multiple users.
+    Users and data must remain segregated at all times.
+    For example, when using physical compute resources, all sensitive data could be securely wiped before another user is given access to that same node.
```
Is wording on could be strong enough here, if we are saying it's mandatory

jemrobinson replied:
- Keep users and data separated (this is mandatory)
- All sensitive data could be securely wiped (one possible way of achieving the above)

JimMadge commented:
> There are definitions in WP3. I'll reproduce them here:
> ...

I suppose those work. Do they cover all possible capability components?

If we use them in this document I think we need to be clear about when we use them (_i.e._ say "Application Component", ideally with a hyperlink to the decomposition definitions, rather than just "Application") and include the definitions in the preamble.

manics commented:
We've got a glossary which defines these terms
https://satre-specification.readthedocs.io/en/latest/glossary.html

A nice (future!) improvement would be to automatically cross-link terms to the glossary definition:
- https://www.sphinx-doc.org/en/master/usage/restructuredtext/directives.html#directive-glossary
- https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#role-term

jemrobinson commented:
Merging as most comments have been addressed and this is quite a big reorganisation of one of the pillars that is holding up other work.
